---
title: "Proxy GxE MR on breastfeeding and cognitive traits"
output:
  html_document:
    toc: yes
    number_sections: yes
bibliography: references.bib
---

```{r setup, include=FALSE}
# Packages
library(data.table)
library(ggplot2)
library(table1)
library(dplyr)
library(modelsummary)
library(forcats)

# Load data
data <- fread("../../temp/merged/data.tsv", stringsAsFactors = TRUE, na.strings = "")
```

# Introduction

This RMarkdown document reports the preliminary results for the proxy gene-by-environment Mendelian randomisation of the influence of breastfeeding on educational attainment on the UK Biobank.

We use a subsample of the UK Biobank, including participants with White British ancestry, no aneuploidy, no mismatch between reported gender and generically-derived sex, no high heterozigosity and SNP missingness, and who did not withdrew their consent in participating (N = 486,400).

To model the data, we use the following fields:

-   Year of birth (field 34);

-   Sex (field 31 for self-reported sex, 22001 for genetically derived sex);

-   Breastfed as a baby or not (field 1677);

-   Genetic principal components (field 22009);

-   Fluid intelligence (fields 20191 and 20016);

-   Age of completion of full time education (field 845);

-   Indices of deprivation:

    -   Townsend deprivation index (field 189);

    -   Indices of multiple deprivation:

        -   England (field 26410);

        -   Scotland (field 26427);

        -   Wales (field 26426);

-   Metabolomics:

    -   Total DHA (field 23450);

    -   DHA/total fatty acids percentage (field 23457);

    -   QC for DHA (field 26231);

-   Genetics QC:

    -   Batch (field 22000);

    -   Plate (field 22004);

    -   Well (field 22008).

# Preparing the data

-   **IQ**: To generate a simple measure of cognitive functions, we use the unweighted sum of correct answers to the fluid intelligence battery of questions, taking mean of the instances of when it was assessed;

```{r cleaning-FI, echo=FALSE, warning=FALSE}
data_clean <-
  data %>%
  # Generate variable for IQ based on the average score for FI
  # HACK this assumes no habituation effect
  mutate(FI_all = rowMeans(across(starts_with("FI")),
                           na.rm = TRUE))
```

-   **Educational attainment**: we derive a measure of educational attainment as the number of years of education required to attain the highest degree, according to their ICSED level @Carter2022;

    | Qualifications (as reported in UK Biobank)              |  ISCED   | Years of education |
    |--------------------------------------------|:---------:|:----------------:|
    | College or University degree                            |    5     |         20         |
    | NVQ or HND or HNC or equivalent                         |    5     |         19         |
    | Other professioal qualification e.g.: nursing, teaching |    4     |         15         |
    | A levels/AS levels or equivalent                        |    3     |         13         |
    | O levels/GCSEs or equivalent                            |    2     |         10         |
    | CSEs or equivalent                                      |    2     |         10         |
    | None of the above                                       |    1     |         7          |
    | Prefer not to answer                                    | Excluded |         NA         |

```{r cleaning-EA, echo=FALSE, warning=FALSE}
# Map degrees and years of education
year_equivalence <-
  c(
    "College or University degree" = 20,
    "NVQ or HND or HNC or equivalent" = 19,
    "Other professional qualifications eg: nursing, teaching" = 15,
    "A levels/AS levels or equivalent" = 13,
    "CSEs or equivalent" = 10,
    "O levels/GCSEs or equivalent" = 10,
    "None of the above" = 7,
    "Prefer not to answer" = NA
  )

# Convert highest qualification to years of education
data_clean <-
  data_clean %>%
  select(starts_with("qualifications")) %>%
  # Take only the max of the years of schooling reported, according to the map
  # Returns -Inf if they're all NA (and prints a lot of warnings)
  transmute(years_of_schooling = apply(., 1, \(x) max(year_equivalence[x], na.rm = TRUE))) %>%
  bind_cols(data_clean) %>%
  # Replace -Inf with NA 
  mutate(years_of_schooling = replace(years_of_schooling, years_of_schooling == -Inf, NA)) %>%
  select(-starts_with("qualifications"))
```

```{r factor-levels-order, echo=FALSE}
## Reorder factors
# Non-forcats solution
# data_clean$genetic_sex <- factor(data_for_table$genetic_sex,
#                        levels = rev(levels(data_for_table$genetic_sex))) # Test
# data_clean$breastfed_as_baby <- factor(data_for_table$breastfed_as_baby,
#                                            levels = c("Yes",
#                                                       "No",
#                                                       "Do not know",
#                                                       "Prefer not to answer"))

# Forcats solution
data_clean <-
  data_clean %>%
  mutate(genetic_sex = fct_rev(genetic_sex), # To test if it works
         breastfed_as_baby = fct_relevel(breastfed_as_baby,
                                         "Yes",
                                         "No",
                                         "Do not know",
                                         "Prefer not to answer"))
```

# Table 1

```{r table1, echo=FALSE, out.width="100%"}
data_for_table <-
  # Round dosage to allelic count
  data_clean %>%
  mutate(rs1535_A = factor(round(rs1535_A)),
         rs174575_C = factor(round(rs174575_C)),
         rs174583_C = factor(round(rs174583_C)))

# Assign alleles for each allelic count
data_for_table <- 
  data_for_table %>%
  mutate(
    rs1535_A = case_when(rs1535_A == 0 ~ "GG",
                         rs1535_A == 1 ~ "AG",
                         rs1535_A == 2 ~ "AA"),
    rs174575_C = case_when(rs174575_C == 0 ~ "GG",
                           rs174575_C == 1 ~ "CG",
                           rs174575_C == 2 ~ "CC"),
    rs174583_C = case_when(rs174583_C == 0 ~ "TT",
                           rs174583_C == 1 ~ "CT",
                           rs174583_C == 2 ~ "CC")
    )

# Create table labels
table_lables <- 
  list(variables = list(genetic_sex="Sex",
                        age="Year of birth",
                        breastfed_as_baby="Breastfed as a baby",
                        FI_all = "FI score",
                        years_of_schooling = "Years in education",
                        Townsend_deprivation_index = "Townsend deprivation index"),
       groups = list("", "rs1535", "rs174575", "rs174583")
  )
strata <- c(list(Total=data_for_table),
            split(data_for_table, data_for_table$rs1535_A),
            split(data_for_table, data_for_table$rs174575_C),
            split(data_for_table, data_for_table$rs174583_C))

table1(strata, table_lables, groupspan = c(1, 3, 3, 3), data = data_for_table)
# TODO fix width with table1 %>% t1flex() %>% flextable::...
```

# Modeling

For the purpose of modelling, we merge together nonresponders and participants who responded "Do not know" or "Prefer not to answer".

```{r recode-missing, echo=FALSE}
data_for_models <-
  data_clean %>%
  mutate(breastfed_as_baby = fct_explicit_na(breastfed_as_baby,
                                             na_level = "Missing"),
         breastfed_as_baby = fct_collapse(breastfed_as_baby,
                                          Breastfed = c("Yes"),
                                          Bottlefed = c("No"),
                                          Missing = c("Missing", "Do not know", "Prefer not to answer")) %>% fct_relevel("Breastfed", "Bottlefed", "Missing"),
         sex = genetic_sex)
```

```{r model-table-setup, echo=FALSE}
# Table arguments: estimate, CI
est_table <- "{estimate} [{conf.low}, {conf.high}]"
stats_table <- "p = {p.value}"

model_table <- 
  \(model) modelsummary(model,
             estimate = est_table,
             statistic = stats_table,
             coef_rename = TRUE,
             # NOTE works on Linux only!
             mc.cores = parallel::detectCores(),
             title = paste("Models for", 
                           # FIXME better to use just the predictor name
                           model[[1]] %>% 
                             broom::tidy() %>%
                             # NOTE: SNP must always be on 1st position for this
                             # to work
                             .[2,1] %>%
                             stringi::stri_extract(regex = "^rs[0-9]+")
                           )
             )
# List of variants 
variants <- c("rs1535_A", "rs174575_C", "rs174583_C")
```

------------------------------------------------------------------------

**NOTE:** the analysis includes complete cases only. The number of observations for each model varies according to the missingness in the predictor and outcome variables

------------------------------------------------------------------------

## Phenotypic associations

Breastfeeding is coded as 1 = breastfed and 0 = not-breastfed. The correlation coefficients are computed using pairwise complete observations.

```{r phenotypic-associations, echo = FALSE}

corr_matrix <-
  data_for_models %>%
  mutate(breastfed_as_baby = recode(breastfed_as_baby, 
                                    "Breastfed" = 1,
                                    "Bottlefed" = 0,
                                    .default = NA_real_)) %>%
  select(breastfed_as_baby, age, Townsend_deprivation_index, FI_all, rs1535_A, rs174575_C, rs174583_C, years_of_schooling, IMD_England) %>%
  rename(breastfed = breastfed_as_baby,
         TDI = Townsend_deprivation_index,
         FI = FI_all,
         rs1535 = rs1535_A,
         rs174575 = rs174575_C,
         rs174583 = rs174583_C) %>%
  cor(use = "pairwise.complete.obs")
  

ggcorrplot::ggcorrplot(corr_matrix,
                       hc.order = TRUE,
                       type = "lower",
                       lab = TRUE,
                       lab_size = 3,
                       outline.color = "white")

```

## Simple model

In the simplest case, we fit $Y \sim SNP$:

```{r simple-models-setup, echo=FALSE}
# Simple model with no interactions: outcome ~ SNP
# Returns a list of lm() objects
simple_model <- function(data,
                         outcome, # Name of the outcome variable
                         SNP_name){ # SNP name in the form rsid_ea
  lapply(split(data, data$breastfed_as_baby),
         \(x) lm(formula(paste(outcome, 
                   "~",
                   SNP_name,
                   "")),
             data = x)
  )
}
```

### Fluid intelligence

```{r simple-model-FI, echo=FALSE}
# HACK using a loop does not print the table, so I need to call them manually
 simple_model(data_for_models, "FI_all", variants[1]) %>%
  model_table()

 simple_model(data_for_models, "FI_all", variants[2]) %>%
  model_table()

 simple_model(data_for_models, "FI_all", variants[3]) %>%
  model_table()
```

### Educational attainment

```{r simple-model-EA, echo=FALSE}
 simple_model(data_for_models, "years_of_schooling", variants[1]) %>%
  model_table()

 simple_model(data_for_models, "years_of_schooling", variants[2]) %>%
  model_table()

 simple_model(data_for_models, "years_of_schooling", variants[3]) %>%
  model_table()
```


## Age and sex as covariates

$Y \sim SNP + age + sex$:

```{r cov-models-setup, echo=FALSE}
# Model with some covariates: outcome ~ SNP + sex + age
# Returns a list of lm() objects
cov_model <- function(data,
                         outcome, # Name of the outcome variable
                         SNP_name){ # SNP name in the form rsid_ea
  lapply(split(data, data$breastfed_as_baby),
         \(x) lm(formula(paste(outcome, 
                   "~",
                   SNP_name,
                   "+ sex + age")),
             data = x)
  )
}
```

### Fluid intelligence

```{r cov-model-FI, echo=FALSE}
# HACK using a loop does not print the table, so I need to call them manually
 cov_model(data_for_models, "FI_all", variants[1]) %>%
  model_table()

 cov_model(data_for_models, "FI_all", variants[2]) %>%
  model_table()

 cov_model(data_for_models, "FI_all", variants[3]) %>%
  model_table()
```

### Educational attainment

```{r cov-model-EA, echo=FALSE}
 cov_model(data_for_models, "years_of_schooling", variants[1]) %>%
  model_table()

 cov_model(data_for_models, "years_of_schooling", variants[2]) %>%
  model_table()

 cov_model(data_for_models, "years_of_schooling", variants[3]) %>%
  model_table()
```


## Model including measures of socioeconomic status

Modelled as $Y \sim SNP + age + sex + SES$, similar to @Caspi2007 (without interaction term)

```{r ses-models-setup, echo=FALSE}
# Model including the TDI as a measure of deprivation
# Returns a list of lm() objects
ses_model <- function(data,
                         outcome, # Name of the outcome variable
                         SNP_name){ # SNP name in the form rsid_ea
  lapply(split(data, data$breastfed_as_baby),
         \(x) lm(formula(paste(outcome, 
                   "~",
                   SNP_name,
                   "+ sex + age + Townsend_deprivation_index")),
             data = x)
  )
}
```

### Fluid intelligence

```{r ses-model-FI, echo=FALSE}
 ses_model(data_for_models, "FI_all", variants[1]) %>%
  model_table()

 ses_model(data_for_models, "FI_all", variants[2]) %>%
  model_table()

 ses_model(data_for_models, "FI_all", variants[3]) %>%
  model_table()
```

### Educational attainment

```{r ses-model-EA, echo=FALSE}
 ses_model(data_for_models, "years_of_schooling", variants[1]) %>%
  model_table()

 ses_model(data_for_models, "years_of_schooling", variants[2]) %>%
  model_table()

 ses_model(data_for_models, "years_of_schooling", variants[3]) %>%
  model_table()
```

## Principal components models

Modelled as $Y \sim SNP + age + SES + \sum^{10}_{n=1}PC_{n}$:

```{r pc-ses-models-setup, echo=FALSE}
# Model including both SES and the first 10 PCs
# Returns a list of lm() objects
pc_ses_model <- function(data,
                         outcome, # Name of the outcome variable
                         SNP_name){ # SNP name in the form rsid_ea
  lapply(split(data, data$breastfed_as_baby),
         \(x) lm(formula(paste(outcome, 
                   "~",
                   SNP_name,
                   "+ sex + age + Townsend_deprivation_index +",
                   paste(paste0("principal_component_", 1:10),
                         collapse = " + "))),
             data = x)
  )
}
```

### Fluid intelligence

```{r pc-ses-model-FI, echo=FALSE}
 pc_ses_model(data_for_models, "FI_all", variants[1]) %>%
  model_table()

 pc_ses_model(data_for_models, "FI_all", variants[2]) %>%
  model_table()

 pc_ses_model(data_for_models, "FI_all", variants[3]) %>%
  model_table()
```

### Educational attainment

```{r pc-ses-model-EA, echo=FALSE}
 pc_ses_model(data_for_models, "years_of_schooling", variants[1]) %>%
  model_table()

 pc_ses_model(data_for_models, "years_of_schooling", variants[2]) %>%
  model_table()

 pc_ses_model(data_for_models, "years_of_schooling", variants[3]) %>%
  model_table()
```

## Genetic QC effect models

Modelled as $Y \sim SNP + age + SES + batch$:

```{r qc-pc-ses-models-setup, echo=FALSE}
# Model including batch
# Returns a list of lm() objects
qc_ses_model <- function(data,
                         outcome, # Name of the outcome variable
                         SNP_name){ # SNP name in the form rsid_ea
  lapply(split(data, data$breastfed_as_baby),
         \(x) lm(formula(paste(outcome, 
                   "~",
                   SNP_name,
                   "+ sex + age + Townsend_deprivation_index + batch")),
             data = x)
  )
}
```

### Fluid intelligence

```{r qc-ses-model-FI, echo=FALSE}
 qc_pc_ses_model(data_for_models, "FI_all", variants[1]) %>%
  model_table()

 qc_pc_ses_model(data_for_models, "FI_all", variants[2]) %>%
  model_table()

 qc_pc_ses_model(data_for_models, "FI_all", variants[3]) %>%
  model_table()
```

### Educational attainment

```{r qc-ses-model-EA, echo=FALSE}
 qc_pc_ses_model(data_for_models, "years_of_schooling", variants[1]) %>%
  model_table()

 qc_pc_ses_model(data_for_models, "years_of_schooling", variants[2]) %>%
  model_table()

 qc_pc_ses_model(data_for_models, "years_of_schooling", variants[3]) %>%
  model_table()
```

# References
