---
title: "Proxy GxE MR on breastfeeding and cognitive traits"
date: "`r Sys.Date()`"
output:
  rmdformats::downcute:
    number_sections: yes
bibliography: references.bib
---

```{r setup, include=FALSE}
# Packages
library(data.table)
library(ggplot2)
library(table1)
library(dplyr)
library(modelsummary)
library(forcats)
library(lubridate)
library(forestplot)

# Load data
data <- fread("../../temp/merged/data.tsv", stringsAsFactors = TRUE, na.strings = "")

unrelateds <- scan("../../temp/merged/unrelateds.txt")

```

```{r basic-cleaning-data, echo=FALSE, warning=FALSE}
data_clean <-
  data %>%
  rename(yob = age,
         mob = `52.0.0`,
         date_FI_online = `20135.0.0`) %>%
  rename_with(~gsub("21003", "age_FI_ac", .x, fixed = TRUE),
              .cols = starts_with("21003")) %>%
  mutate(date_of_birth = ym(paste(yob, mob)),
            date_FI_online = ymd_hms(date_FI_online),
            # NOTE the following works just by sorcery alone--don't understand the logic
         FI_age_online = interval(start = date_of_birth, 
                                  end = date_FI_online)/years(1)) %>%
# remove unrelated individuals
  filter(!(eid %in% unrelateds)) %>%
  select(-c("mob", "date_FI_online", "date_of_birth"))

rm(data)

# FIXME something is generating too many NAs!
# NOTE FI_age_ac is a misnomer, the age is for when attended assessment
# centre. I correct it midway through but this whole code should be
# restructured
fi_score_long <-
  data_clean %>%
  select(eid, starts_with("FI_centre")) %>%
  tidyr::pivot_longer(cols = starts_with("FI_centre"),
                      names_to = "FI_centre_instance",
                      values_to = "FI_score")

fi_age_long <- 
   data_clean %>%
   select(eid, starts_with("age_FI_ac")) %>%
   tidyr::pivot_longer(starts_with("age_FI_ac"),
                       names_to = "FI_centre_instance",
                       values_to = "FI_age_ac") %>%
   mutate(FI_centre_instance = case_when(
     FI_centre_instance == "age_FI_ac.0.0" ~ "FI_centre_0_",
     FI_centre_instance == "age_FI_ac.1.0" ~ "FI_centre_1_",
     FI_centre_instance == "age_FI_ac.2.0" ~ "FI_centre_2_",
     FI_centre_instance == "age_FI_ac.3.0" ~ "FI_centre_3_",
                                  )) 
# get table with age at each instance and fi
fi_ac <-
full_join(fi_score_long, fi_age_long, by = c("eid", "FI_centre_instance")) %>%
  # Make NA where no info on FI for that instance
  mutate(FI_age_ac = case_when(
    is.na(FI_score) ~ NA_real_,
    !is.na(FI_score) ~ as.double(FI_age_ac))) %>%
  group_by(eid) %>%
  slice_min(order_by = FI_age_ac, with_ties = FALSE) %>%
  ungroup() %>%
  rename(FI_ac_score = FI_score)

data_clean <-
  full_join(fi_ac, data_clean, by = "eid") %>%
  # Take first instance (= youngest)
  mutate(
    FI_all = case_when(
      FI_age_ac > FI_age_online ~ as.integer(FI_online),
      FI_age_ac < FI_age_online ~ as.integer(FI_ac_score),
      FI_age_ac == FI_age_online ~ as.integer(FI_ac_score),
      TRUE ~ NA_integer_
      ),
    # Take the earliest instance of age (same principle)
    # rounding the age in both to keep it consistent
    age_at_FI_assessment = case_when(
      FI_age_ac > FI_age_online ~ as.integer(round(FI_age_online)),
      FI_age_ac < FI_age_online ~ as.integer(round(FI_age_ac)),
      FI_age_ac == FI_age_online ~ as.integer(round(FI_age_ac)),

      TRUE ~ NA_integer_
      ),
    # Finally, create a factor for which score of intelligence has been used
    # according to the same rules
    FI_scale = case_when(
      FI_age_ac > FI_age_online ~ "Online",
      FI_age_ac < FI_age_online ~ "Assessment centre",
      TRUE ~ NA_character_
      )
    )
```

# Introduction

This RMarkdown document reports the results of a proxy gene-by-environment Mendelian randomisation analysis of the influence of breastfeeding on educational attainment on the UK Biobank.

We use a subsample of the UK Biobank, including participants with White British ancestry, no aneuploidy, no mismatch between reported gender and generically-derived sex, no high heterozigosity and SNP missingness, and who did not withdrew their consent in participating, and without any known relatedness in the sample (N = 486,400).

To model the data, we use the following fields:

-   Year of birth (field 34);

-   Sex (field 31 for self-reported sex, 22001 for genetically-derived sex);

-   Breastfed as a baby or not (field 1677);

-   Genetic principal components (field 22009);

-   Fluid intelligence (fields 20191 and 20016);

-   Age of completion of full time education (field 845);

-   Indices of deprivation:

    -   Townsend deprivation index (field 189);

    -   Indices of multiple deprivation:

        -   England (field 26410);

        -   Scotland (field 26427);

        -   Wales (field 26426);

-   Metabolomics:

    -   Total DHA (field 23450);

    -   DHA/total fatty acids percentage (field 23457);

    -   QC for DHA (field 26231);

-   Genetics QC:

    -   Batch (field 22000);

    -   Plate (field 22004);

    -   Well (field 22008).

# Preparing the data

-   **IQ**: we used the earliest available instance for either the online questionnaire or the assessment centre score, using age at completion and which scale is it coming from as a covariate; we expect the variable on the scale of origin to have an approximate effect of 1 (the online questionnaire had a question more in the assessment);

-   ***Breastfeeding***: coded as the reported breastfeeding status at the first instance;

-   **Educational attainment**: we derive a measure of educational attainment as the number of years of education required to attain the highest degree, according to their ICSED level @Carter2022;

    | Qualifications (as reported in UK Biobank)              |  ISCED   | Years of education |
    |---------------------------------------------------------|:--------:|:------------------:|
    | College or University degree                            |    5     |         20         |
    | NVQ or HND or HNC or equivalent                         |    5     |         19         |
    | Other professioal qualification e.g.: nursing, teaching |    4     |         15         |
    | A levels/AS levels or equivalent                        |    3     |         13         |
    | O levels/GCSEs or equivalent                            |    2     |         10         |
    | CSEs or equivalent                                      |    2     |         10         |
    | None of the above                                       |    1     |         7          |
    | Prefer not to answer                                    | Excluded |         NA         |

```{r cleaning-EA, echo=FALSE, warning=FALSE}
# Map degrees and years of education
year_equivalence <-
  c(
    "College or University degree" = 20,
    "NVQ or HND or HNC or equivalent" = 19,
    "Other professional qualifications eg: nursing, teaching" = 15,
    "A levels/AS levels or equivalent" = 13,
    "CSEs or equivalent" = 10,
    "O levels/GCSEs or equivalent" = 10,
    "None of the above" = 7,
    "Prefer not to answer" = NA
  )

# Convert highest qualification to years of education
data_clean <-
  data_clean %>%
  select(starts_with("qualifications")) %>%
  # Take only the max of the years of schooling reported, according to the map
  # Returns -Inf if they're all NA (and prints a lot of warnings)
  transmute(years_of_schooling = apply(., 1, \(x) max(year_equivalence[x], na.rm = TRUE))) %>%
  bind_cols(data_clean) %>%
  # Replace -Inf with NA 
  mutate(years_of_schooling = replace(years_of_schooling, years_of_schooling == -Inf, NA)) %>%
  select(-starts_with("qualifications"))
```

-   **Age**: we use the year of birth for educational attainment, or age at completion for the IQ score;

```{r factor-levels-order, echo=FALSE}
## Reorder factors
# Non-forcats solution
# data_clean$genetic_sex <- factor(data_for_table$genetic_sex,
#                        levels = rev(levels(data_for_table$genetic_sex))) # Test
# data_clean$breastfed_as_baby <- factor(data_for_table$breastfed_as_baby,
#                                            levels = c("Yes",
#                                                       "No",
#                                                       "Do not know",
#                                                       "Prefer not to answer"))

# Forcats solution
data_clean <-
  data_clean %>%
  mutate(genetic_sex = fct_rev(genetic_sex), # To test if it works
         breastfed_as_baby = fct_relevel(breastfed_as_baby,
                                         "Yes",
                                         "No",
                                         "Do not know",
                                         "Prefer not to answer"))
```

# Table 1

```{r table1, echo=FALSE, out.width="100%"}
data_for_table <-
  # Round dosage to allelic count
  data_clean %>%
  mutate(rs1535_A = factor(round(rs1535_A)),
         rs174575_C = factor(round(rs174575_C)),
         rs174583_C = factor(round(rs174583_C)))

# Assign alleles for each allelic count
data_for_table <- 
  data_for_table %>%
  mutate(
    rs1535_A = case_when(rs1535_A == 0 ~ "GG",
                         rs1535_A == 1 ~ "AG",
                         rs1535_A == 2 ~ "AA"),
    rs174575_C = case_when(rs174575_C == 0 ~ "GG",
                           rs174575_C == 1 ~ "CG",
                           rs174575_C == 2 ~ "CC"),
    rs174583_C = case_when(rs174583_C == 0 ~ "TT",
                           rs174583_C == 1 ~ "CT",
                           rs174583_C == 2 ~ "CC")
    )

# Create table labels
table_lables <- 
  list(variables = list(genetic_sex="Sex",
                        yob="Year of birth",
                        breastfed_as_baby="Breastfed as a baby",
                        FI_ac_score = "FI score (assessment centre)",
                        FI_online = "FI score (online follow-up)",
                        FI_all = "FI score (first available instance from any source)",
                        age_at_FI_assessment = "Age at first available instance for FI)",
                        years_of_schooling = "Years in full-time education",
                        Townsend_deprivation_index = "Townsend deprivation index"),
       groups = list("", "rs1535", "rs174575", "rs174583")
  )
strata <- c(list(Total=data_for_table),
            split(data_for_table, data_for_table$rs1535_A),
            split(data_for_table, data_for_table$rs174575_C),
            split(data_for_table, data_for_table$rs174583_C))

table1(strata, table_lables, groupspan = c(1, 3, 3, 3), data = data_for_table)
# TODO fix width with table1 %>% t1flex() %>% flextable::...

# Recover a bit of memory
rm(strata, data_for_table)
```

# Analysis

For the purpose of modelling, we merge together nonresponders and participants who responded "Do not know" or "Prefer not to answer".

```{r recode-missing, echo=FALSE}
data_for_models <-
  data_clean %>%
  mutate(breastfed_as_baby = fct_explicit_na(breastfed_as_baby,
                                             na_level = "Missing"),
         breastfed_as_baby = fct_collapse(breastfed_as_baby,
                                          Breastfed = c("Yes"),
                                          Bottlefed = c("No"),
                                          Missing = c("Missing", "Do not know", "Prefer not to answer")) %>% fct_relevel("Breastfed", "Bottlefed", "Missing"),
         sex = genetic_sex)
```

```{r model-table-setup, echo=FALSE}
# Table arguments: estimate, CI
est_table <- "{estimate} [{conf.low}, {conf.high}]"
stats_table <- "p = {p.value}"

model_table <- 
  \(model) modelsummary(model,
             estimate = est_table,
             statistic = stats_table,
             coef_rename = TRUE,
             # NOTE works on Linux only!
             mc.cores = parallel::detectCores(),
             title = paste("Models for", 
                           # FIXME better to use just the predictor name
                           model[[1]] %>% 
                             broom::tidy() %>%
                             # NOTE: SNP must always be on 1st position for this
                             # to work
                             .[2,1] %>%
                             stringi::stri_extract(regex = "^rs[0-9]+")
                           )
             )
# List of variants 
variants <- c("rs1535_A", "rs174575_C", "rs174583_C")
```

------------------------------------------------------------------------

**NOTE:** the analysis includes complete cases only. The number of observations for each model can vary according to the missingness in the predictor and outcome variables

------------------------------------------------------------------------

## Phenotypic associations

Breastfeeding is coded as 1 = breastfed and 0 = not-breastfed. The correlation coefficients are computed using pairwise complete observations.

```{r phenotypic-associations, echo = FALSE}

corr_matrix <-
  data_for_models %>%
  mutate(breastfed_as_baby = recode(breastfed_as_baby, 
                                    "Breastfed" = 1,
                                    "Bottlefed" = 0,
                                    .default = NA_real_)) %>%
  select(breastfed_as_baby, yob, Townsend_deprivation_index, FI_all, FI_age_ac, FI_age_online, FI_ac_score, FI_online, rs1535_A, rs174575_C, rs174583_C, years_of_schooling, IMD_England, age_at_FI_assessment) %>%
  rename(breastfed = breastfed_as_baby,
         TDI = Townsend_deprivation_index,
         FI_all_instances = FI_all,
         Age_at_FI_assessment_centre = FI_age_ac,
         Age_at_FI_online = FI_age_online,
         FI_assessment_centre = FI_ac_score,
         FI_online = FI_online,
         year_of_birth = yob,
         rs1535 = rs1535_A,
         rs174575 = rs174575_C,
         rs174583 = rs174583_C) %>%
  cor(use = "pairwise.complete.obs")
  

ggcorrplot::ggcorrplot(corr_matrix,
                       hc.order = TRUE,
                       type = "lower",
                       lab = TRUE,
                       lab_size = 2,
                       outline.color = "white")

```

### Breastfeeding and IQ

There is evidence from the previous literature that breastfeeding might be associated with IQ @Hartwig2019. To test this hypothesis, we fit $Y \sim breastfeeding + sex + age + TDI + \sum^{10}_{n=1}PC_{n}$ , where $breastfeeding$ is coded as the reference level:

```{r breastfeeding-iq, echo = FALSE}

formula_bf_iq_agesex <- formula(paste("FI_all ~ breastfed_as_baby + sex + age_at_FI_assessment + FI_scale"))


model_bf_iq_agesex <- 
  data_for_models %>%
  lm(formula_bf_iq_agesex, data =.)

summary(model_bf_iq_agesex)
```

```{r breastfeeding-iq-tdi, echo = FALSE}

formula_bf_iq_tdi <- formula(paste("FI_all ~ breastfed_as_baby + sex + age_at_FI_assessment + Townsend_deprivation_index"))


model_bf_iq_tdi <- 
  data_for_models %>%
  lm(formula_bf_iq_tdi, data =.)

summary(model_bf_iq_tdi)
```

```{r breastfeeding-iq-pcs, echo = FALSE}

formula_bf_iq_tdi_pcs <- formula(paste("FI_all ~ breastfed_as_baby + sex + age_at_FI_assessment + Townsend_deprivation_index +",
                             paste(paste0("principal_component_", 1:10),
                                   collapse = " + ")
                             ))


model_bf_iq_tdi_pcs <- 
  data_for_models %>%
  lm(formula_bf_iq_tdi_pcs, data =.)

summary(model_bf_iq_tdi_pcs)

```

```{r breastfeeding-ea-tdi, echo = FALSE}

formula_bf_ea_tdi <- formula(paste("years_of_schooling ~ breastfed_as_baby + sex + yob + Townsend_deprivation_index"))


model_bf_ea_tdi <- 
  data_for_models %>%
  lm(formula_bf_ea_tdi, data =.)

summary(model_bf_ea_tdi)
```

```{r breastfeeding-ea-tdi-pcs, echo = FALSE}

formula_bf_ea_tdi_pcs <- formula(paste("years_of_schooling ~ breastfed_as_baby + sex + yob + Townsend_deprivation_index +",
                             paste(paste0("principal_component_", 1:10),
                                   collapse = " + ")
                             ))


model_bf_ea_tdi_pcs <- 
  data_for_models %>%
  lm(formula_bf_ea_tdi_pcs, data =.)

summary(model_bf_ea_tdi_pcs)

```

## FADS and breastfeeding

We also want to check whether FADS and breastfeeding are associated. They should not, as the change in allele dosage should not be different between breastfed and bottlefed individuals.

```{r breastfeeding-rs1535, echo=FALSE}

breastfeeding_rs1535 <-
  formula(paste("breastfed_as_baby ~ rs1535_A +",
                             paste(paste0("principal_component_", 1:10),
                                   collapse = " + ")
                             ))

data_for_models %>%
  mutate(breastfed_as_baby = na_if(breastfed_as_baby, "Missing")) %>%
  glm(breastfeeding_rs1535, family = binomial, data = .) %>%
  summary()
```

```{r breastfeeding-rs174575, echo=FALSE}

breastfeeding_rs174575 <-
  formula(paste("breastfed_as_baby ~ rs174575_C +",
                             paste(paste0("principal_component_", 1:10),
                                   collapse = " + ")
                             ))

data_for_models %>%
  mutate(breastfed_as_baby = na_if(breastfed_as_baby, "Missing")) %>%
  glm(breastfeeding_rs174575, family = binomial, data = .) %>%
  summary()
```

```{r breastfeeding-rs174583, echo=FALSE}

breastfeeding_rs174583 <-
  formula(paste("breastfed_as_baby ~ rs174583_C +",
                             paste(paste0("principal_component_", 1:10),
                                   collapse = " + ")
                             ))

data_for_models %>%
  mutate(breastfed_as_baby = na_if(breastfed_as_baby, "Missing")) %>%
  glm(breastfeeding_rs174583, family = binomial, data = .) %>%
  summary()
```

## Replicating @Caspi2007's analysis

We attempt to replicate the model shown in @Caspi2007 as closely as possible by fitting $Y \sim SNP + breasfeeding + SNP \times breastfeeding + SES$.

Breastfeeding is coded once again as 1 = breastfed and 0 = bottlefed.

```{r caspi-analysis-rs1535-no-agesex, echo=FALSE}

model_caspi_rs1535_noagesex <- 
  data_for_models %>%
  mutate(breastfed_as_baby = na_if(breastfed_as_baby, "Missing"),
         breastfed_as_baby = recode(breastfed_as_baby, "Breastfed" = 1,
                      "Bottlefed" = 0)) %>%
  lm(FI_all ~ rs1535_A + breastfed_as_baby + rs1535_A*breastfed_as_baby + Townsend_deprivation_index + FI_scale + age_at_FI_assessment, data =.)

summary(model_caspi_rs1535_noagesex)

```

```{r caspi-analysis-rs174575-no-agesex, echo=FALSE}

model_caspi_rs174575_noagesex <- 
  data_for_models %>%
  mutate(breastfed_as_baby = na_if(breastfed_as_baby, "Missing"),
         breastfed_as_baby = recode(breastfed_as_baby, "Breastfed" = 1,
                      "Bottlefed" = 0)) %>%
  lm(FI_all ~ rs174575_C + breastfed_as_baby + rs174575_C*breastfed_as_baby + Townsend_deprivation_index + FI_scale + age_at_FI_assessment, data =.)

summary(model_caspi_rs174575_noagesex)

```

@Caspi2007 's analysis was done using a single cohort with the same age and balanced for sex, so we also try to repeat the model adjusting for age and sex.

```{r caspi-analysis-rs1535-agesex, echo=FALSE}

model_caspi_rs1535_agesex <- 
  data_for_models %>%
  mutate(breastfed_as_baby = na_if(breastfed_as_baby, "Missing"),
         breastfed_as_baby = recode(breastfed_as_baby, "Breastfed" = 1,
                      "Bottlefed" = 0)) %>%
  lm(FI_all ~ rs1535_A + breastfed_as_baby + rs1535_A*breastfed_as_baby + Townsend_deprivation_index + sex + yob, data =.)

summary(model_caspi_rs1535_agesex)
```

```{r caspi-analysis-rs174575-agesex, echo=FALSE}

model_caspi_rs174575_agesex <- 
  data_for_models %>%
  mutate(breastfed_as_baby = na_if(breastfed_as_baby, "Missing"),
         breastfed_as_baby = recode(breastfed_as_baby, "Breastfed" = 1,
                      "Bottlefed" = 0)) %>%
  lm(FI_online ~ rs174575_C + breastfed_as_baby + rs174575_C*breastfed_as_baby + Townsend_deprivation_index + sex + yob, data =.)

summary(model_caspi_rs174575_agesex)
```

There are a few important differences between UK Biobank and the studies in @Caspi2007 that might also explain the results:

-   Standardised childhood IQ test (Dunedin and E-risk) vs unstandardised adult IQ test (UK Biobank);

-   Measure of individual SES based on parental occupation (Dunedin) vs regional deprivation (UK Biobank);

-   Cohort homogeneous for age and with repeated IQ measurements (Dunedin and E-risk) vs first-instance questions at different ages (UK Biobank).

## SNP association model with no covariates

In the simplest case, we fit $Y \sim SNP$:

```{r simple-models-setup, echo=FALSE}
# Simple model with no interactions: outcome ~ SNP
# Returns a list of lm() objects
simple_model_fi <- function(data,
                         outcome, # Name of the outcome variable
                         SNP_name){ # SNP name in the form rsid_ea
  lapply(split(data, data$breastfed_as_baby),
         \(x) lm(formula(paste(outcome, 
                   "~",
                   SNP_name,
                   "+ FI_scale")),
             data = x) %>%
           broom::tidy()
  ) %>%
    bind_rows(.id = "stratum")
}

simple_model_ea <- function(data,
                         outcome, # Name of the outcome variable
                         SNP_name){ # SNP name in the form rsid_ea
  lapply(split(data, data$breastfed_as_baby),
         \(x) lm(formula(paste(outcome, 
                   "~",
                   SNP_name,
                   "")),
             data = x) %>%
           broom::tidy()
  ) %>%
    bind_rows(.id = "stratum")
}
```

```{r forest-plot, echo=FALSE}
draw_forest_plot <- function(model_table){
 model_table %>%
  filter(!(term == "(Intercept)")) %>%
  mutate(
    mean = estimate,
         lower= estimate-(1.96*std.error),
         upper= estimate+(1.96*std.error)) %>%
  select(stratum, term, mean, std.error, lower, upper, p.value) %>%
 group_by(stratum) %>%
  forestplot(
    labeltext = term,
    boxsize = .1,
    grid = TRUE,
    ci.vertices = TRUE,
    ci.vertices.height = 0.05,
  ) %>%
  fp_add_lines("steelblue") %>%
  fp_add_header("Estimate") %>%
  fp_set_style(box = c("darkblue", "darkred", "black") %>% lapply(function(x) gpar(fill = x, col = "#555555")),
               line = c("darkblue", "darkred", "black"),
               default = gpar(vertices = TRUE)) %>%
    fp_set_zebra_style("#EFEFEF")
}

```

### Fluid intelligence

```{r simple-model-FI, echo=FALSE}
# HACK using a loop does not print the table, so I need to call them manually
 simple_model_fi(data_for_models, "FI_all", variants[1]) %>%
  draw_forest_plot()

 simple_model_fi(data_for_models, "FI_all", variants[2]) %>%
  draw_forest_plot()

 simple_model_fi(data_for_models, "FI_all", variants[3]) %>%
  draw_forest_plot()
```

### Educational attainment

```{r simple-model-EA, echo=FALSE}
 simple_model_ea(data_for_models, "years_of_schooling", variants[1]) %>%
  draw_forest_plot()

 simple_model_ea(data_for_models, "years_of_schooling", variants[2]) %>%
  draw_forest_plot()

 simple_model_ea(data_for_models, "years_of_schooling", variants[3]) %>%
  draw_forest_plot()
```

## Age and sex as covariates

$Y \sim SNP + age + sex$:

```{r cov-models-setup, echo=FALSE}
# Model with some covariates: outcome ~ SNP + sex + age
# Returns a list of lm() objects

cov_model <- function(data,
                         outcome, # Name of the outcome variable
                         SNP_name){ # SNP name in the form rsid_ea
  lapply(split(data, data$breastfed_as_baby),
         \(x) lm(formula(paste(outcome, 
                   "~",
                   SNP_name,
                   "+ sex + age_at_FI_assessment + FI_scale")),
             data = x) 
  )
}

cov_model_fi <- function(data,
                         outcome, # Name of the outcome variable
                         SNP_name){ # SNP name in the form rsid_ea
  lapply(split(data, data$breastfed_as_baby),
         \(x) lm(formula(paste(outcome, 
                   "~",
                   SNP_name,
                   "+ sex + age_at_FI_assessment + FI_scale")),
             data = x) %>%
           broom::tidy()
  ) %>%
    bind_rows(.id = "stratum")
}


cov_model_ea <- function(data,
                         outcome, # Name of the outcome variable
                         SNP_name){ # SNP name in the form rsid_ea
  lapply(split(data, data$breastfed_as_baby),
         \(x) lm(formula(paste(outcome, 
                   "~",
                   SNP_name,
                   "+ sex + yob")),
             data = x) %>%
           broom::tidy()
  ) %>%
    bind_rows(.id = "stratum")
}
```

### Fluid intelligence

```{r cov-model-FI, echo=FALSE}
# HACK using a loop does not print the table, so I need to call them manually
 cov_model_fi(data_for_models, "FI_all", variants[1]) %>%
  draw_forest_plot()

 cov_model_fi(data_for_models, "FI_all", variants[2]) %>%
  draw_forest_plot()

 cov_model_fi(data_for_models, "FI_all", variants[3]) %>%
  draw_forest_plot()
```

### Educational attainment

```{r cov-model-EA, echo=FALSE}
 cov_model_ea(data_for_models, "years_of_schooling", variants[1]) %>%
  draw_forest_plot()

 cov_model_ea(data_for_models, "years_of_schooling", variants[2]) %>%
  draw_forest_plot()

 cov_model_ea(data_for_models, "years_of_schooling", variants[3]) %>%
  draw_forest_plot()
```

## Principal components models

Modelled as $Y \sim SNP + sex + age + \sum^{10}_{n=1}PC_{n}$:

```{r pc-models-setup, echo=FALSE}
# Model including first 10 PCs
# Returns a list of lm() objects
pc_model_fi <- function(data,
                         outcome, # Name of the outcome variable
                         SNP_name){ # SNP name in the form rsid_ea
  lapply(split(data, data$breastfed_as_baby),
         \(x) lm(formula(paste(outcome, 
                   "~",
                   SNP_name,
                   "+ sex + age_at_FI_assessment + FI_scale + ",
                   paste(paste0("principal_component_", 1:10),
                         collapse = " + "))),
             data = x) %>%
           broom::tidy()
  ) %>%
    bind_rows(.id = "stratum")
}

pc_model_ea <- function(data,
                         outcome, # Name of the outcome variable
                         SNP_name){ # SNP name in the form rsid_ea
  lapply(split(data, data$breastfed_as_baby),
         \(x) lm(formula(paste(outcome, 
                   "~",
                   SNP_name,
                   "+ sex + yob + ",
                   paste(paste0("principal_component_", 1:10),
                         collapse = " + "))),
             data = x) %>%
           broom::tidy()
  ) %>%
    bind_rows(.id = "stratum")
}

```

### Fluid intelligence

```{r pc-ses-model-FI, echo=FALSE}
 pc_model_fi(data_for_models, "FI_all", variants[1]) %>%
  draw_forest_plot()

 pc_model_fi(data_for_models, "FI_all", variants[2]) %>%
  draw_forest_plot()

 pc_model_fi(data_for_models, "FI_all", variants[3]) %>%
  draw_forest_plot()
```

### Educational attainment

```{r pc-ses-model-EA, echo=FALSE}
 pc_model_ea(data_for_models, "years_of_schooling", variants[1]) %>%
  draw_forest_plot()

 pc_model_ea(data_for_models, "years_of_schooling", variants[2]) %>%
  draw_forest_plot()

 pc_model_ea(data_for_models, "years_of_schooling", variants[3]) %>%
  draw_forest_plot()
```

# References
